src/Core/Geometry/IndexedGeometry.hpp

```diff
@@ -0,0 +1,491 @@
+    inline PointCloudIndexLayer() : GeometryIndexLayer( staticSemanticName ) {}
+    /// \brief Generate linearly spaced indices with same size as \p attr vertex buffer
+    void generateIndicesFromAttributes( const AttribArrayGeometry& attr );
+
+    static constexpr const char* staticSemanticName = "PointCloud";
+
+  protected:
+    template <class... SemanticNames>
+    inline PointCloudIndexLayer( SemanticNames... names ) :
+        GeometryIndexLayer( staticSemanticName, names... ) {}
+};
+
+struct RA_CORE_API TriangleIndexLayer : public GeometryIndexLayer<Vector3ui> {
+    inline TriangleIndexLayer() : GeometryIndexLayer( staticSemanticName ) {}
+
+    static constexpr const char* staticSemanticName = "TriangleMesh";
+
+  protected:
+    template <class... SemanticNames>
+    inline TriangleIndexLayer( SemanticNames... names ) :
+        GeometryIndexLayer( staticSemanticName, names... ) {}
+};
+
+class RA_CORE_API PolyIndexLayer : public GeometryIndexLayer<VectorNui>
+{
+    inline PolyIndexLayer() : GeometryIndexLayer( "PolyMesh" ) {}
+
+  protected:
+    template <class... SemanticNames>
+    inline PolyIndexLayer( SemanticNames... names ) : GeometryIndexLayer( "PolyMesh", names... ) {}
+};
+
+class RA_CORE_API LineIndexLayer : public GeometryIndexLayer<Vector2ui>
+{
+    inline LineIndexLayer() : GeometryIndexLayer( "LineMesh" ) {}
+
+  protected:
+    template <class... SemanticNames>
+    inline LineIndexLayer( SemanticNames... names ) : GeometryIndexLayer( "LineMesh", names... ) {}
+};
+
+#ifdef MULTI_INDEX_MIMIC_TRIANGLE_MESH
+const MultiIndexedGeometry::IndexContainerType& MultiIndexedGeometry::getIndices() const {
+    const auto& abstractLayer = getFirstLayerOccurrence( TriangleIndexLayer::staticSemanticName );
+    return static_cast<const TriangleIndexLayer&>( abstractLayer ).collection();
+}
+
+MultiIndexedGeometry::IndexContainerType& MultiIndexedGeometry::getIndicesWithLock() {
+    auto& abstractLayer = getFirstLayerOccurrenceWithLock( TriangleIndexLayer::staticSemanticName );
+    return static_cast<TriangleIndexLayer&>( abstractLayer ).collection();
+}
+
+void MultiIndexedGeometry::indicesUnlock() {
+    unlockFirstLayerOccurrence( TriangleIndexLayer::staticSemanticName );
+}
+
+void MultiIndexedGeometry::setIndices( IndexContainerType&& indices ) {
+
+    auto l           = std::make_unique<TriangleIndexLayer>();
+    LayerKeyType key = std::make_pair( l->semantics(), "" );
+
+    auto it = m_indices.find( key );
+    CORE_ASSERT( it != m_indices.end(), "TriangleIndex should be inserted already" );
+
+    static_cast<TriangleIndexLayer&>( *( it->second.second ) ).collection() = std::move( indices );
+
+    notify();
+}
+
+class RA_CORE_API TriangleMesh : public MultiIndexedGeometry
+{
+  public:
+    inline TriangleMesh() : MultiIndexedGeometry() {
+        auto l = std::make_unique<TriangleIndexLayer>();
+        addLayer( std::move( l ) );
```

Comment:
to be more efficient, less verbose and simplify future maintenance, use simply 
```.cpp
addLayer( {} );
```
rvalues and temporaries are the same ...

Reply:
No, your change breaks everything: addLayer takes a generic type as input, you need to define the specialized type by calling its constructor (`TriangleIndexLayer`).
This is why the tests were crashing.